Sources/SPMTestSupport/Toolchain.swift

```diff
@@ -58,3 +58,38 @@ extension UserToolchain {
         }
     }
 }
+
+extension UserToolchain {
+    /// Helper function to determine if async await actually works in the current environment.
+    public func supportsSwiftConcurrency() -> Bool {
+      #if os(macOS)
+        if #available(macOS 12.0, *) {
+            // On macOS 12 and later, concurrency is assumed to work.
+            return true
+        }
+        else {
+            // On macOS 11 and earlier, we don't know if concurrency actually works because not all SDKs and toolchains have the right bits.  We could examine the SDK and the various libraries, but the most accurate test is to just try to compile and run a snippet of code that requires async/await support.  It doesn't have to actually do anything, it's enough that all the libraries can be found (but because the library reference is weak we do need the linkage reference to `_swift_task_create` and the like).
+            do {
+                try testWithTemporaryDirectory { tmpPath in
+                    let inputPath = tmpPath.appending(component: "foo.swift")
+                    try localFileSystem.writeFileContents(inputPath, string: "public func foo() async {}\nTask { await foo() }")
+                    let outputPath = tmpPath.appending(component: "foo")
+                    let toolchainPath = self.swiftCompilerPath.parentDirectory.parentDirectory
+                    let backDeploymentLibPath = toolchainPath.appending(components: "lib", "swift-5.5", "macosx")
+                    try Process.checkNonZeroExit(arguments: ["/usr/bin/xcrun", "--toolchain", toolchainPath.pathString, "swiftc", inputPath.pathString, "-Xlinker", "-rpath", "-Xlinker", backDeploymentLibPath.pathString, "-o", outputPath.pathString])
+                    try Process.checkNonZeroExit(arguments: [outputPath.pathString])
```

Comment:
This was the most accurate test I could come up with that checks whether concurrency actually works, given the various combinations of toolchain and SDK that had various bugs.  This adds about a second per skipped test to the running time, and only on macOS 11 — I couldn't find a good way to memoize this result based on the toolchain instance, since extensions can't add properties and different toolchains could have different results — I guess we could have a mapping from toolchain to result, but I'm not sure it's worth it.

Reply:
But the check itself seems robust.